packages/astro/src/build.ts

```diff
@@ -212,16 +212,21 @@ ${stack}
 
     // write to disk and free up memory
     timer.write = performance.now();
-    await Promise.all(
-      Object.keys(buildState).map(async (id) => {
-        const outPath = new URL(`.${id}`, astroConfig.dist);
-        const parentDir = path.dirname(fileURLToPath(outPath));
-        await fs.promises.mkdir(parentDir, { recursive: true });
-        await fs.promises.writeFile(outPath, buildState[id].contents, buildState[id].encoding);
-        delete buildState[id];
-        delete depTree[id];
-      })
-    );
+    for (const id of Object.keys(buildState)) {
+      const outPath = new URL(`.${id}`, astroConfig.dist);
+      const parentDir = path.dirname(fileURLToPath(outPath));
+      await fs.promises.mkdir(parentDir, {recursive: true});
+      const handle = await fs.promises.open(outPath, "w")
+      await fs.promises.writeFile(handle, buildState[id].contents, buildState[id].encoding);
+
+      // Ensure the file handle is not left hanging which will
+      // result in the garbage collector loggin errors in the console
+      // when it eventually has to close them.
+      await handle.close();
```

Comment:
would just adding this change work?

Reply:
I am not sure. I have a gut feeling that a writeFile() call might take long enough that it allows more and more async jobs (from Promise.all) to start and open a file handle. I will check, though.

Reply:
Tried with this and got the same "too many open files" error. I think how it works is that pretty much each individual promise is sort of advancing the same pace so they all arrive at the writeFile() call at the same time.
```
await Promise.all(Object.keys(buildState).map(async (id) => {
      const outPath = new URL(`.${id}`, astroConfig.dist);
      const parentDir = path.dirname(fileURLToPath(outPath));
      await fs.promises.mkdir(parentDir, {recursive: true});
      const handle = await fs.promises.open(outPath, "w")
      await fs.promises.writeFile(handle, buildState[id].contents, buildState[id].encoding);

      // Ensure the file handle is not left hanging
      await handle.close();

      delete buildState[id];
      delete depTree[id];
    }));
```

Reply:
Maybe instead of making it serial, we can use a promise pool? https://futurestud.io/tutorials/node-js-run-async-functions-in-batches

basically we limit the # of concurrent promises running at the same time

Reply:
I thought about that but came in the result that it would be overengineering at least for the moment. I don't think there necessarily is a benefit in having concurrent writes happening on the disk and if there is. I'd say that's a enhancement while this just fixes the immediate bug? (Also I have a dependency aversion.) :) 

I'm happy to add that feature in a later PR if it is needed!

Reply:
Yeah I agree, this seems like the right way to do it, I doubt writing files is going to be a performance bottleneck, so doing it serial should be fine.